core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.underfs.options.ListOptions;
+import alluxio.underfs.options.MkdirsOptions;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Preconditions;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A base abstract {@link UnderFileSystem}.
+ */
+@ThreadSafe
+public abstract class BaseUnderFileSystem implements UnderFileSystem {
+  /** The UFS {@link AlluxioURI} used to create this {@link BaseUnderFileSystem}. */
+  protected final AlluxioURI mUri;
+
+  /** A map of property names to values. */
+  protected HashMap<String, String> mProperties = new HashMap<>();
+
+  /**
+   * Constructs an {@link BaseUnderFileSystem}.
+   *
+   * @param uri the {@link AlluxioURI} used to create this ufs
+   */
+  protected BaseUnderFileSystem(AlluxioURI uri) {
+    Preconditions.checkNotNull(uri);
+    mUri = uri;
+  }
+
+  @Override
+  public void configureProperties() throws IOException {
+    // Default implementation does not update any properties.
+  }
+
+  @Override
+  public OutputStream create(String path) throws IOException {
+    return create(path, CreateOptions.defaults());
+  }
+
+  @Override
+  public boolean deleteDirectory(String path) throws IOException {
+    return deleteDirectory(path, DeleteOptions.defaults());
+  }
+
+  @Override
+  public Map<String, String> getProperties() {
+    return Collections.unmodifiableMap(mProperties);
+  }
+
+  @Override
+  public UnderFileStatus[] listStatus(String path, ListOptions options) throws IOException {
+    if (!options.isRecursive()) {
+      return listStatus(path);
+    }
+    // Clean the path by creating a URI and turning it back to a string
+    AlluxioURI uri = new AlluxioURI(path);
+    path = uri.toString();
+    List<UnderFileStatus> returnPaths = new ArrayList<>();
+    Queue<UnderFileStatus> pathsToProcess = new ArrayDeque<>();
```

Comment:
I did not right this code, but I think the double ended-ness comes at no extra cost. A LinkedList is an alternative but ArrayDeque seems to be preferred because of line caching.